roms_tools/setup/cdr_forcing.py

```diff
@@ -21,171 +18,210 @@
 
     @model_validator(mode="after")
-    def check_tracer_concentrations(self) -> "Release":
-        """Ensure tracer concentractions match length of 'times'."""
+    def check_unique_name(self) -> "ReleaseCollector":
+        """Check that all releases have unique names."""
+        names = [release.name for release in self.releases]
+        duplicates = [name for name, count in Counter(names).items() if count > 1]
 
-        num_times = len(self.times)
+        if duplicates:
+            raise ValueError(
+                f"Multiple releases share the following name(s): {', '.join(repr(d) for d in duplicates)}. "
+                "Each release must have a unique name."
+            )
 
-        for tracer, val in self.tracer_concentrations.items():
+        return self
 
-            if isinstance(val, list):
-                if len(val) != num_times:
-                    raise ValueError(
-                        f"The length of tracer '{tracer}' ({len(val)}) does not match the number of times ({num_times})."
-                    )
+    def determine_release_type(self) -> "ReleaseCollector":
+        """Ensure all releases are of the same type, and set the release_type."""
+        if not all(
+            isinstance(release, type(self.releases[0])) for release in self.releases
+        ):
+            raise ValueError("Not all releases are of the same type.")
 
-        return self
+        return type(self.releases[0])
 
-    @model_validator(mode="after")
-    def extend_to_endpoints(self) -> "Release":
-        """Ensure that the release time series starts at `start_time` and ends at
-        `end_time`.
 
-        If `volume_fluxes` is a list and does not cover the endpoints, zero volume fluxes are added.
-        Tracer concentrations are extended accordingly by duplicating endpoint values.
+class CDRForcingDatasetBuilder:
+    """Constructs the xarray `Dataset` to be saved as NetCDF."""
 
-        If `times` is empty, [start_time, end_time] is used with zero fluxes and duplicated tracer values.
-        """
-        if not self.times:
-            self.times = [self.start_time, self.end_time]
-            self.volume_fluxes = [self.volume_fluxes, self.volume_fluxes]
-            self.tracer_concentrations = {
-                tracer: [vals, vals]
-                if not isinstance(vals, list)
-                else [vals[0], vals[0]]
-                for tracer, vals in self.tracer_concentrations.items()
+    def __init__(self, releases, model_reference_date, release_type):
+        self.releases = releases
+        self.model_reference_date = model_reference_date
+        self.release_type = release_type
+
+    def build(self) -> xr.Dataset:
+        all_times = [np.array(r.times, dtype="datetime64[ns]") for r in self.releases]
+        union_times = np.unique(np.concatenate(all_times))
+        union_rel_times = [
+            convert_to_relative_days(t, self.model_reference_date) for t in union_times
+        ]
```

Comment:
Can we vectorize the relative time conversion?

```
ref = np.datetime64(model_reference_date, "ns")
day_delta = np.timedelta64(1, "D")

relative = union_times - ref
relative_days = relative / day_delta
```

This should be much faster since it acts on the entire array at once instead of converting each value individually

Reply:
Good point. Done in 44ed95d.